src/ParseCLP.js

```diff
@@ -0,0 +1,898 @@
+/**
+ * Copyright (c) 2015-present, Parse, LLC.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @flow
+ */
+
+import ParseRole from './ParseRole';
+import ParseUser from './ParseUser';
+
+type UsersMap = { [userId: string]: boolean | any };
+type PermissionsMap = { [permission: string]: UsersMap };
+
+const PUBLIC_KEY = '*';
+
+const VALID_PERMISSIONS: Map<string, UsersMap> = new Map<string, UsersMap>();
+VALID_PERMISSIONS.set('get', {});
+VALID_PERMISSIONS.set('find', {});
+VALID_PERMISSIONS.set('count', {});
+VALID_PERMISSIONS.set('create', {});
+VALID_PERMISSIONS.set('update', {});
+VALID_PERMISSIONS.set('delete', {});
+VALID_PERMISSIONS.set('addField', {});
+
+const VALID_PERMISSIONS_EXTENDED: Map<string, UsersMap> = new Map<string, UsersMap>();
+VALID_PERMISSIONS_EXTENDED.set('protectedFields', []);
+
+/**
+ * Creates a new CLP.
+ * If no argument is given, the CLP has no permissions for anyone.
+ * If the argument is a Parse.User, the CLP will have read and write
+ *   permission for only that user.
+ * If the argument is any other JSON object, that object will be interpretted
+ *   as a serialized CLP created with toJSON().
+ *
+ * <p>A CLP, or Class Level Permissions can be added to any
+ * <code>Parse.Schema</code> to restrict access to only a subset of users
+ * of your application.</p>
+ * @alias Parse.CLP
+ */
+class ParseCLP {
+  permissionsMap: PermissionsMap;
+
+  /**
+   * @param {(Parse.User|Parse.Role|Object)} userId The user to initialize the CLP for
+   */
+  constructor(userId: ParseUser | ParseRole | PermissionsMap) {
```

Comment:
Why not `permissionId`?

Reply:
Because the Parse.CLP has the intention to give a permission to a user/group. The permissions are always: find, get, count, create, update, delete, addField

Reply:
Because the Parse.CLP has the intention to give a permission to a user/group. The permissions are always: find, get, count, create, update, delete, addField

Reply:
I believe that will be call 99% with correct parameters, so I do not want waste time doing this validation every call.